arch/x86/kvm/vmx/pkvm/hyp/iommu.c

```diff
@@ -393,25 +395,75 @@ static bool sync_shadow_pasid_dir_entry(struct pgt_sync_data *sdata)
 	return false;
 }
 
+static struct pkvm_ptdev *iommu_find_ptdev(struct pkvm_iommu *iommu, u16 bdf, u32 pasid)
+{
+	struct pkvm_ptdev *p;
+
+	list_for_each_entry(p, &iommu->ptdev_head, iommu_node) {
+		if (match_ptdev(p, bdf, pasid))
+			return p;
+	}
+
+	return NULL;
+}
+
+static struct pkvm_ptdev *iommu_add_ptdev(struct pkvm_iommu *iommu, u16 bdf, u32 pasid)
+{
+	struct pkvm_ptdev *ptdev = pkvm_get_ptdev(bdf, pasid);
+
+	if (!ptdev)
+		return NULL;
+
+	list_add_tail(&ptdev->iommu_node, &iommu->ptdev_head);
+	return ptdev;
+}
+
+static void iommu_del_ptdev(struct pkvm_iommu *iommu, struct pkvm_ptdev *ptdev)
+{
+	list_del_init(&ptdev->iommu_node);
+	pkvm_put_ptdev(ptdev);
+}
+
 /* sync pasid table entry when guest_ptep valid, otherwise un-present it */
 static bool sync_shadow_pasid_table_entry(struct pgt_sync_data *sdata)
 {
+	u16 bdf = sdata->vaddr >> DEVFN_SHIFT;
+	u32 pasid = sdata->vaddr & ((1UL << MAX_NR_PASID_BITS) - 1);
+	struct pkvm_iommu *iommu = pgt_to_pkvm_iommu(sdata->spgt);
+	struct pkvm_ptdev *ptdev = iommu_find_ptdev(iommu, bdf, pasid);
 	struct pasid_entry *shadow_pte = sdata->shadow_ptep, tmp_pte = {0};
 	struct pasid_entry *guest_pte;
+	bool synced = false;
 	u64 type, aw;
 
+	if (!ptdev) {
+		ptdev = iommu_add_ptdev(iommu, bdf, pasid);
```

Comment:
We add ptdev here even if we are going to remove it below in `if (!sdata->guest_ptep)`. In particular, if guest_ptep is null and the shadow pte is not present either, this newly added ptdev will not be removed, so it will be around in iommu->ptdev_head despite not really used. Won't it cause any issues (e.g. iommu_audit_did() misfailure)?

It seems it would be better to move this iommu_add_ptdev logic (lines 439-443) after `if (!sdata->guest_ptep)` (after line 460) and put the iommu_del_ptdev logic (lines 453-455) under `if (ptdev)` check.

Reply:
The current log doesn't have issue because the caller actually guarantees this case won't happen. If both !sdata->guest_pte and shadow_pte not present, the caller (sync_shadow_id_cb and free_shadow_id_cb) will not call iommu_id_sync_entry. But changing the logic to what you describe is also fine to me.

Reply:
Yeah, indeed.